src/Mod/Sketcher/Gui/TaskSketcherElements.cpp

```diff
@@ -677,128 +971,82 @@ void TaskSketcherElements::leaveEvent (QEvent * event)
-                                                                (tr("Elliptical Arc") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-                                                                (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-                                                                (QString::fromLatin1("%1-").arg(i)+tr("Elliptical Arc")))   :
-        type == Part::GeomArcOfHyperbola::getClassTypeId()    ? ( isNamingBoxChecked ?
-                                                                (tr("Hyperbolic Arc") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-                                                                (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-                                                                (QString::fromLatin1("%1-").arg(i)+tr("Hyperbolic Arc")))   :
-        type == Part::GeomArcOfParabola::getClassTypeId()    ? ( isNamingBoxChecked ?
-                                                                (tr("Parabolic Arc") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-                                                                (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-                                                                (QString::fromLatin1("%1-").arg(i)+tr("Parabolic Arc")))   :
-        type == Part::GeomBSplineCurve::getClassTypeId()    ? ( isNamingBoxChecked ?
-                                                                (tr("BSpline") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-                                                                (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-                                                                (QString::fromLatin1("%1-").arg(i)+tr("BSpline")))   :
-        ( isNamingBoxChecked ?
-          (tr("Other") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-          (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-          (QString::fromLatin1("%1-").arg(i)+tr("Other"))),
-        i-1,
-        sketchView->getSketchObject()->getVertexIndexGeoPos(i-1,Sketcher::PointPos::start),
-        sketchView->getSketchObject()->getVertexIndexGeoPos(i-1,Sketcher::PointPos::mid),
-        sketchView->getSketchObject()->getVertexIndexGeoPos(i-1,Sketcher::PointPos::end),
-        type,
-        construction,
-        false));
-
-        setItemVisibility(i-1, mode);
+        Base::Type type = (*it)->getTypeId();
+        bool construction = Sketcher::GeometryFacade::getConstruction(*it);
+
+        ElementWidgetIcons::Type icontype = construction ? ElementWidgetIcons::Type::Construction : ElementWidgetIcons::Type::Normal;
+
+
+        QListWidgetItem* itemN = new QListWidgetItem;
+
+        ElementData* elementData = new ElementData(i - 1,

```

Comment:
I think this will leak, as  I do not follow where this dynamic memory is released.
ElementItemDelegate is a Q_OBJECT, but it is created based on the specific constructor, so it has no parent that could eventually delete it.

QVariant does not call the destructor for underlying pointers.

See also:
https://forum.qt.io/topic/65270/qvariant-pointer-data-after-deletion-of-qstandarditem-qvariant-data-set-with-setdata

It appears there are different solutions. As far as I read, it is best to use a derived class of QListWidgetItem (as it was in the original code), which is light and has in-built memory management. If that route cannot be used for the delegation (which I do not see why not, but I am new to this), then we need to either provide it with a parent, or create a QListWidgetItem derived class, and in the destructor of the QListWidgetItem, we need to delete the pointer to the Element data there.

Is there a compelling reason why we need to use ElementData instead of having ElementItem (derived from QListWidgetItem), as it was before this PR?

Reply:
Because I couldn't find how to access ElementItem from const QModelIndex& index in ElementItemDelegate::editorEvent

So I could not access the item data as  Ididn't know how to get ElementItem from QModelIndex.

Reply:
To be clearer, I couldn't find how to do : 
`itemData->clickedOn = ElementType::edge;`
using the available
`QModelIndex& index`
I can't find how to call itemFromIndex function from QListWidget from the delegate.

I drive me nuts, then I found about this setData and other examples of custom delegate all used that so I thought it was the proper way of doing it.

Reply:
Ok I found a possibility if we want to use ElementItem subclass instead of ElementData:

When we create the delegate, we can pass the pointer to ElementView as an argument. Then we can access the ElementItem from the index using that pointer.

Do you think this is a better way of doing things?

Reply:
I have some learning to do before I can weigh in.... but thanks for looking into alternatives. I will come back to you.

Reply:
Ok.
Besides, you mentioned before that you want to add filter for type 'internal aligment' (bspline circles, ellipse lines...). If you can tell me how to check if geometry is of that type, I can add that filter.

Reply:
It is done via GeometryFacade. Now there is a member function isInternalAligned. But then you need to construct the facade yourself. We could use a static version of that function.

I have pushed a commit to the branch. So you can now just do:
```
GeometryFacade::isInternalAligned(geo);
```
if you happen to have a geometry facade object already created, then you can do with less overhead:
```
gf->isInternalAligned();
```

Reply:
Ok. I have been reflecting on the overall problem.

There is no good place to release ElementData. One could then make ElementData not a pointer and pass by value. This I think it would be very heavy. The best possibility to maintain ElementData, is probably to use both the item and the element data and control the lifetime via the item. However, this looks convoluted as we are mixing two different ways of doing almost the same thing.

Your alternative of passing the view pointer to the delegate is what is used in ExpressionDelegate in TaskSketcherConstraints. Basically, you already have the explicit constructor, so in addition to passing it to the parent, you may well keep a copy. If you can get what you need from the view, then I think we should go this path.

Is it possible for you to do these changes?

Reply:
I wonder if the overall problem with that is not that data is stored in the item rather than in the model...
Also as all this mainly turns around selection, Qt has the possibility to implement a `QItemSelectionModel` which may be more appropriate.

Reply:
Regarding the internal geo filter. What about icon color? What hue magic number to use ? Here I wrote 200 totally randomly.

```
                if (alp > 127 && hue >= 0) {
                    if (sat > 127 && (hue > 330 || hue < 30)) {
                        clr.setHsv((hue + 240) % 360, sat, val, alp);
                        imgConstr.setPixelColor(ix, iy, clr);
                        clr.setHsv((hue + 300) % 360, sat, val, alp);
                        imgExt.setPixelColor(ix, iy, clr);
                        clr.setHsv((hue + 200) % 360, sat, val, alp);
                        imgInt.setPixelColor(ix, iy, clr);
                    }
                    else if (sat < 64 && val > 192)
                    {
                        clr.setHsv(240, (255-sat), val, alp);
                        imgConstr.setPixel(ix, iy, clr.rgba());
                        clr.setHsv(300, (255-sat), val, alp);
                        imgExt.setPixel(ix, iy, clr.rgba());
                        clr.setHsv(200, (255-sat), val, alp);
                        imgInt.setPixel(ix, iy, clr.rgba());
                    }
                }
```

Reply:
@0penBrain 

It sounds good to me, but I do not have experience with it. I have taken a look to QItemSelectionModel and I have done an Internet search but I have not found something that I could directly apply. I guess that if we go that route, we would need to ask somebody who knows.

I am willing to look into different proposals.

@PaddleStroke 

Internal Alignment color is:
SbColor DrawingParameters::InternalAlignedGeoColor                  (0.7f,0.7f,0.5f);     // #B2B27F -> (178,178,127)

I am not very sensitive to colour. @0penBrain was the one implementing those functions (if I remember correctly). He may be in a better position to help you tweak the number.

Reply:
These function should be make flexible to match user-chosen color but that's another talk. :)
According default color for internal alignments, I would say hue of `hue+60` and val of `(int)val*2/3`
So apply hue of `60` in the 2nd if block.
I can't test know but you can try and add a screenshot for checking.

Reply:
if (sat > 127 && (hue > 330 || hue < 30)) {
                        clr.setHsv((hue + 240) % 360, sat, val, alp);
                        imgConstr.setPixelColor(ix, iy, clr);
                        clr.setHsv((hue + 300) % 360, sat, val, alp);
                        imgExt.setPixelColor(ix, iy, clr);
                        clr.setHsv((hue + 60) % 360, sat, (int) val * 2 / 3, alp);
                        imgInt.setPixelColor(ix, iy, clr);
                    }
                    else if (sat < 64 && val > 192)
                    {
                        clr.setHsv(240, (255-sat), val, alp);
                        imgConstr.setPixel(ix, iy, clr.rgba());
                        clr.setHsv(300, (255-sat), val, alp);
                        imgExt.setPixel(ix, iy, clr.rgba());
                        clr.setHsv(60, (255-sat), val, alp);
                        imgInt.setPixel(ix, iy, clr.rgba());
                    }
![image](https://user-images.githubusercontent.com/19984177/196471394-bfc652f1-0385-4ce1-99d5-c32b8f788e25.png)

Reply:
Ok I see now. However now that all icons are visible, this is no longer needed, so I removed the color modification for the red points for constr & external. Internal points do change color

Reply:
I'm not sure this is good idea ATM. It doesn't hurt and should IMO be handled later, time for users to feedback about if and how they would like this to change. ;) Just my opinion.

Reply:
Oh right. This is likely to be worthy of a 15 page of very important discussion on the forum.

IMO it hurts because it's confusing that the icon doesn't match the view. As they say, you don't put a picture of white beans on a can of black beans.

I will let Abdullah decide on this.

Reply:
An icon is just an indication. It has to be not confusing, but it doesn't have to be an exact representation too. ;)

Reply:
Which is true, but the icons have too many colors, it's just confusing. 
I think the best is that we simplify the icons by changing the color of dots from red to white. (ie white point and white line, unless it's the selected element). So that construction is all blue and external all pink. This will simplify the icons, making them much clearer. And outlining the selected element.

Reply:
Like this : 
![image](https://user-images.githubusercontent.com/19984177/196486014-cde8757d-531c-4e93-9356-cb53accdd0fb.png)

Reply:
Here what it looks like 
![image](https://user-images.githubusercontent.com/19984177/196487407-12513c8e-e2db-4645-8578-b61a32a39d21.png)

Reply:
Here what it looks like for circles and lines
![image](https://user-images.githubusercontent.com/19984177/196487617-5f1544a3-06ea-467a-a2e0-93a0834474fb.png)

Reply:
Do whatever you want (here too). :) But IMO you're not learning from the past. Riding too much horses at the same time.

Reply:
@0penBrain 

I have been checking use of QItemSelectionModel in other parts of FreeCAD. Nothing that I see goes beyond choosing from different default strategies (basically which flags to use). On the other side, I do understand that a substantial part of the issue is about selection, and that, if it were to be handled from the selection model, it would be a better way.

Here I must apologise as I do not know better. Then I must ask: Are you willing at this time to give it a shot yourself?

If the answer is yes, I will wait for it. If the answer is no, then I will advice to revert to the Item subclass as a way forward. Please let me know.

Reply:
@abdullahtahiriyo Sorry, not enough spare time ATM to dive into this. ;)

Reply:
Ok got it, I'll revert colors to original. After it merges I'll open a discussion regarding the colors to have people thoughts on the subect.